### PR TITLE
feat: speedup test by only check constraints on less rows for MockProver

### PR DESCRIPTION
**Pull request by lispc**
Most of the rows in circuits are "muted" by `q_step * ....`.  So for testing purpose, since we know which rows are "active" for our circuit, we can only test the constraints/lookups on "active" rows.

Speedup by this PR is significant, in my Mac M1, bitwise_gadget_simple test time changes from 640s to 47s. And in github CI, full CI time goes down from 42min to 12min, within which there are 7min compiling time. So the real speedup is about (42-7)/(12-7) = 7 times. If we implement [rust compiling cache](https://github.com/appliedzkp/zkevm-circuits/issues/351#issuecomment-1051442818) later, the CI time can be further reduce to 5min.

This PR is not ready for merge currently. The PR on halo2 side(<https://github.com/appliedzkp/halo2/pull/38>) should be merged first, then I will change the halo2 in Cargo.toml here.

### CODE REVIEW

File: zkevm-circuits/src/evm_circuit/execution.rs
Comment (z2trillion) [inline zkevm-circuits/src/evm_circuit/execution.rs:407]: I don't think you do anything with the returned offset values. Do you need to include this change to the function signatures in this PR?

Comment (lispc) [reply zkevm-circuits/src/evm_circuit/execution.rs:407]: yes you are right . Result<usize, Error> is remain of my first (not successful) version of changes, it is not needed in fact. I will remove it later.

File: Cargo.toml
Comment (CPerezz) [inline Cargo.toml:17]: Once https://github.com/appliedzkp/halo2/pull/38 is merged. I'll release a new tag and let's use it. Revs are really confusing and difficult to follow.

File: zkevm-circuits/src/evm_circuit.rs
Comment (CPerezz) [inline zkevm-circuits/src/evm_circuit.rs:419]: I wonder if it's possible to extract the active rows having a `MockProver` instance and if so, I would implement this functionality for `MockProver` instead.

On that way, we are able to test all of our test circuits on that manner, not only this one. And also we don't need this fn call, and instead we can simply use `prover.verify_at_active_rows`.

I belive it should be possible if we return `active_rows` after calling `MockProver::run`.

**Review by CPerezz — COMMENTED** [review]
Added a comment which can be related to modifications in the halo2 PR.

Let me know what you think :)

File: zkevm-circuits/src/evm_circuit.rs
Comment (lispc) [reply zkevm-circuits/src/evm_circuit.rs:419]: it may not be easy? since selectors like "q_step" are active column, it may be difficult for MockProver to know them as  zero..

Comment (CPerezz) [reply zkevm-circuits/src/evm_circuit.rs:419]: I see. Makes sense! It's sad but I guess it is what we have for now.

**Review by ed255 — COMMENTED** [review]
LGTM!  I'm not submitting "Approve" yet mainly due to the halo2 dependency that must be updated once appliedzkp/halo2#38 is merged as CPerezz has mentioned.
I also think that the comments from z2trillion should be attended as well :)

Thanks a lot for working on this! We will all benefit from faster testing times in the CI :)

**Comment by CPerezz** [comment]
I just merged https://github.com/appliedzkp/halo2/pull/38 and then released a tag for it: https://github.com/appliedzkp/halo2/releases/tag/v2022_03_06

You can now update all Halo2 git deps for the workspace members and this can be merged after @z2trillion comments are addressed/resolved

cc @lispc

**Comment by lispc** [comment]
CI error:

```
 Documenting bitvec v0.22.3
error: couldn't generate documentation: No such file or directory (os error 2)
Error:   |
  = note: failed to create or modify "/home/runner/work/zkevm-circuits/zkevm-circuits/target/doc/bitvec/vec/struct.BitVec.html"

error: could not document `bitvec`
Error: could not document `bitvec`
```

I think it is not related to this PR? maybe some upstream source ( cargo? bitvec? ) changed?

**Comment by ed255** [comment]
> CI error:
> 
> ```
>  Documenting bitvec v0.22.3
> error: couldn't generate documentation: No such file or directory (os error 2)
> Error:   |
>   = note: failed to create or modify "/home/runner/work/zkevm-circuits/zkevm-circuits/target/doc/bitvec/vec/struct.BitVec.html"
> 
> error: could not document `bitvec`
> Error: could not document `bitvec`
> ```
> 
> I think it is not related to this PR? maybe some upstream source ( cargo? bitvec? ) changed?

@CPerezz also encountered this error once, and if I remember correctly, it just went away by rerunning the CI.  Currently we're depending on two different version of `bitvec` (one is pulled by halo2, the other is pulled by ethereum types).  My impression is that the issue appears because both versions use the same path to build the documentation, and the documentations are built in parallel, which may cause a data race between the two versions reading and writing into the same folder.

From the `cargo doc` log:
```
Warning: The lib target `bitvec` in package `bitvec v0.22.3` has the same output filename as the lib target `bitvec` in package `bitvec v0.20.5 ([https://github.com/ed255/bitvec.git?rev=5cfc5fa8496c66872d21905e677120fc3e79693c#5cfc5fa8)`.](https://github.com/ed255/bitvec.git?rev=5cfc5fa8496c66872d21905e677120fc3e79693c#5cfc5fa8)%60.)
```

For now the easiest solution is re-running the CI when this happens.  One of the versions of `bitvec` we're using is a patched fork controlled by us, maybe we can tweak it to use a different doc folder...

**Comment by CPerezz** [comment]
Hi @ed255 @lispc .

As you can see, I re-run all checks and the error disappeared.
The issue I think, is related to a bug in Cargo which sometimes leads to a compilation error for the docs.

I'll file an issue for it later today, but for now don't worry too much. As this will disappear once the patch is not needed anymore.

**Comment by CPerezz** [comment]
I do also think @ed255 that if we use `resolver=2` (since we're not using edition 2021 yet) we could get rid of this with a high probability.

Do you think making a PR for the resolver or the edition change would make sense?

**Review by CPerezz — APPROVED** [review]
LGTM!!

Nice job @lispc :partying_face:

**Pull request merged**
